src/main/java/com/google/cloud/spanner/pgadapter/statements/IntermediateStatement.java

```diff
@@ -95,24 +95,37 @@ protected ParsedStatement replaceKnownUnsupportedQueries(ParsedStatement parsedS
     return parsedStatement;
   }
 
+  /** @return The number of SQL statements in this {@link IntermediateStatement} */
+  public int getStatementCount() {
+    return statements.size();
+  }
+
   /**
-   * Determines the result type based on the given sql string. The sql string must already been
-   * stripped of any comments that might precede the actual sql string.
+   * Determines the statement types based on the given statement strings. The statement string must
+   * already been stripped of any comments that might precede the actual statement string.
    *
-   * @param parsedStatement The parsed statement to determine the type of result for
-   * @return The {@link ResultType} that the given sql string will produce
+   * @param statements The statement strings to determine the type for
+   * @return The list of {@link StatementType} that the given statement strings will produce
    */
-  protected static ResultType determineResultType(ParsedStatement parsedStatement) {
-    if (parsedStatement.isUpdate()) {
-      return ResultType.UPDATE_COUNT;
-    } else if (parsedStatement.isQuery()) {
-      return ResultType.RESULT_SET;
-    } else {
-      return ResultType.NO_RESULT;
+  protected static ImmutableList<StatementType> determineStatementTypes(
+      ImmutableList<String> statements) {
+    ImmutableList.Builder<StatementType> builder = ImmutableList.builder();
+    for (String sql : statements) {
+      if (PARSER.isUpdateStatement(sql)) {
```

Comment:
Done, but doesn't PARSER.parse do the same thing? 
`ParsedStatement parse(Statement statement, QueryOptions defaultQueryOptions) {
    String sql = removeCommentsAndTrim(statement.getSql());
    ClientSideStatementImpl client = parseClientSideStatement(sql);
    if (client != null) {
      return ParsedStatement.clientSideStatement(client, statement, sql);
    } else if (isQuery(sql)) {
      return ParsedStatement.query(statement, sql, defaultQueryOptions);
    } else if (isUpdateStatement(sql)) {
      return ParsedStatement.update(statement, sql);
    } else if (isDdlStatement(sql)) {
      return ParsedStatement.ddl(statement, sql);
    }
    return ParsedStatement.unknown(statement, sql);
  }`

Reply:
Yes, sorry, I should have been a little more elaborate (and also; we will probably try to optimize the `PARSER.parse` method to do that a little more intelligently). The reason that we should call `PARSER.parse(String)` is that it will return a `ParsedStatement`, and that is what we should keep in our list of statements. The reason for that is that:
1. `ParsedStatement` contains helper methods like `isQuery()` and `isUpdate()` that do not need to re-parse the statement. And as we potentially need to know what type of statement it is multiple times, it is better to only parse it once.
2. `ParsedStatement.isQuery()` will also handle client side statements correctly, that is; it knows whether a client side statement is a query or statement without a result.

So we should change the type of `this.statements` from `ImmutableList<String>` to `ImmutableList<ParsedStatement>`.

Reply:
Done.